### PR TITLE
Referrer breakdown in creator dashboard

### PR DESCRIPTION
**Pull request by swoopej**
The last section of the creator dashboard is here!  This is a big pr, but pretty much all of the action is in `ReferrerBreakdownViewHolder` and `ReferrerBreakdownHolderViewModel`, and most of the logic is fairly straightforward - just a lot of outputs.  Feel free to suggest renames, there's a lot of outputs so wanna make it clear what everything is.  The colors and layout will be updated in a final design pass next, so unless there is glaring errors there let's not worry about it.  A gif 4 u:

referrer_breakdown

### CODE REVIEW

File: app/src/main/java/com/kickstarter/ui/views/ReferrerBreakdownView.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  private Double internalSweepAngle;
+  private RectF outerRectangle;
+  private RectF innerRectangle;
+  private Paint paint;
+
+  public ReferrerBreakdownView(final Context context, final AttributeSet attributeSet) {
+    super(context, attributeSet);
+    this.paint = new Paint();
+    this.outerRectangle = new RectF(0, 0 , 0, 0);
+    this.innerRectangle = new RectF(0, 0 , 0, 0);
```
Comment (swoopej) [inline app/src/main/java/com/kickstarter/ui/views/ReferrerBreakdownView.java]: doing this in the intializer because our linter helpfully informed me that `onDraw` can get called lots and lots of times it's bad to allocate things there.

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+import com.kickstarter.models.Project;
+import com.kickstarter.services.apiresponses.ProjectStatsEnvelope;
+import com.kickstarter.ui.viewholders.CreatorDashboardReferrerBreakDownViewHolder;
+
+import java.util.List;
+
+import rx.Observable;
+import rx.subjects.PublishSubject;
+
+public interface CreatorDashboardReferrerBreakdownHolderViewModel {
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: Let's add some javadoc for these `Inputs` and `Outputs` 🙂

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+    public ViewModel(final @NonNull Environment environment) {
+      super(environment);
+
+      final Observable<Project> currentProject = this.projectAndProjectStatsInput
+        .map(PairUtils::first);
+
+      final Observable<List<ProjectStatsEnvelope.ReferrerStats>> referrerStats = this.projectAndProjectStatsInput
+        .map(PairUtils::second)
+        .map(ProjectStatsEnvelope::referralDistribution);
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: ::::::: siq pointfree :::::

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+      final Observable<List<ProjectStatsEnvelope.ReferrerStats>> customReferrers = referrerStats
+        .flatMap(rs ->
+          Observable.from(rs).filter(r -> r.referrerTypeEnumType() == ReferrerType.CUSTOM).toList()
+        );
+
+      this.projectAndAveragePledge = cumulativeStats
+        .map(ProjectStatsEnvelope.CumulativeStats::averagePledge)
+        .map(Float::intValue)
+        .compose(avg -> Observable.combineLatest(currentProject, avg, Pair::create));
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: rather than doing this `.combineLatest` within a `compose`, it'd be more readable to make this `avg` value an intermediate `Observable<Float>` variable, then the whole output would read as:

```java
this.projectAndAveragePledge = Observable.combineLatest(currentProject, averagePledge, Pair::create);
```

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+      this.customReferrerColor = Observable.just(ReferrerColor.CUSTOM.getReferrerColor())
+        .compose(Transformers.takeWhen(this.projectAndProjectStatsInput));
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: can use `takeWhen` as a static import

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+      this.customReferrerPercent = customReferrers
+        .flatMap(rs ->
+          Observable.from(rs)
+            .reduce(0f, (accum, stat) -> accum + stat.percentageOfDollars())
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: niceeeeeeee

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+      this.projectAndInternalReferrerPledgedAmount = Observable.combineLatest(
+        currentProject,
+        this.internalReferrerPledgedAmount,
+        Pair::create
+      );
+
+      this.pledgedViaCustomLayoutIsGone = this.customReferrerPledgedAmount
+        .map(amount -> amount > 0f)
+        .map(BooleanUtils::negate);
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: rather than having a `.map(BooleanUtils::negate);` here and L163, 167, it's probably just more straightforward to change the logic in the `amount` lambda to be something like `amount <= 0f`

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+    private final Observable<Float> internalReferrerPercent;
+    private final Observable<String> internalReferrerPercentText;
+    private final Observable<Float> internalReferrerPledgedAmount;
+    private final Observable<Boolean> pledgedViaCustomLayoutIsGone;
+    private final Observable<Boolean> pledgedViaExternalLayoutIsGone;
+    private final Observable<Boolean> pledgedViaInternalLayoutIsGone;
+    private final Observable<Pair<Project, Float>> projectAndCustomReferrerPledgedAmount;
+    private final Observable<Pair<Project, Float>> projectAndExternalReferrerPledgedAmount;
+    private final Observable<Pair<Project, Float>> projectAndInternalReferrerPledgedAmount;
+
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: lil newline, but siqqqqq that these are all `Observable` outputs!!

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        .map(amount -> amount > 0f)
+        .map(BooleanUtils::negate);
+    }
+
+    public final Inputs inputs = this;
+    public final Outputs outputs = this;
+
+    private final PublishSubject<Pair<Project, ProjectStatsEnvelope>> projectAndProjectStatsInput = PublishSubject.create();
+
+    private final Observable<Pair<Project, Integer>> projectAndAveragePledge;
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: alpha

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+    @Override
+    public @NonNull Observable<Pair<Project, Float>> projectAndInternalReferrerPledgedAmount() {
+      return this.projectAndInternalReferrerPledgedAmount;
+    }
+    @Override
+    public @NonNull Observable<Pair<Project, Integer>> projectAndAveragePledge() {
+      return this.projectAndAveragePledge;
+    }
+    @Override
+    public @NonNull Observable<Integer> customReferrerColor() {
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: can also alpha here

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerStatsHolderViewModel.java [hunk 11]
```diff
@@ -49,7 +49,7 @@ public ViewModel(final @NonNull Environment environment) {
     final private class OrderByBackersReferrerStatsComparator implements Comparator<ProjectStatsEnvelope.ReferrerStats> {
       @Override
       public int compare(final ProjectStatsEnvelope.ReferrerStats o1, final ProjectStatsEnvelope.ReferrerStats o2) {
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerStatsHolderViewModel.java]: not part of this pr but `@NonNull`

File: app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+      android:layout_height="match_parent"
+      android:layout_weight="1"
+      android:gravity="left"
+      android:paddingStart="@dimen/grid_1"
+      style="@style/Title2"
+      tools:text="$54" />
+
+  </LinearLayout>
+
+</LinearLayout>
```
Comment (luoser) [inline app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml]: lil newline

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+
+import rx.observers.TestSubscriber;
+
+public class CreatorDashboardReferrerBreakdownHolderViewModelTest extends KSRobolectricTestCase {
+  private CreatorDashboardReferrerBreakdownHolderViewModel.ViewModel vm;
+
+  private final TestSubscriber<Float> customReferrerPercent = new TestSubscriber<>();
```
Comment (luoser) [inline app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java]: alpha

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+    this.vm.outputs.projectAndAveragePledge().subscribe(this.projectAndAveragePledge);
+    this.vm.outputs.projectAndCustomReferrerPledgedAmount().subscribe(this.projectAndCustomReferrerPledgedAmount);
+    this.vm.outputs.projectAndExternalReferrerPledgedAmount().subscribe(this.projectAndExternalReferrerPledgedAmount);
+    this.vm.outputs.projectAndInternalReferrerPledgedAmount().subscribe(this.projectAndInternalReferrerPledgedAmount);
+  }
+
+  @Test
+  public void testCustomReferrerColor() {
+    final Project project = ProjectFactory.project();
+    final ProjectStatsEnvelope statsEnvelope = ProjectStatsEnvelopeFactory.projectStatsEnvelope();
```
Comment (luoser) [inline app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java]: it doesn't technically matter because the customReferrerColor will always be the same, but we could either add fake custom stats to this envelope to make it clearer that we're testing on custom referrers, OR we could even refactor the viewmodel to only emit these colors when we need them (i.e. takeWhen(customReferrerStats) rather than takeWhen(projectStatsEnvelope))

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+      .percentageOfDollars(55f)
+      .referrerType(ReferrerType.CUSTOM.getReferrerType())
+      .build();
+    final List<ProjectStatsEnvelope.ReferrerStats> referrerList = Arrays.asList(
+      fiftyFivePercentCustomReferrer);
+    final ProjectStatsEnvelope statsEnvelope = ProjectStatsEnvelopeFactory.projectStatsEnvelope().toBuilder().referralDistribution(referrerList).build();
+
+    setUpEnvironment(environment());
+    this.vm.inputs.projectAndStatsInput(Pair.create(project, statsEnvelope));
+    this.customReferrerPercentText.assertValues(NumberUtils.flooredPercentage(55f * 100f));
```
Comment (luoser) [inline app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java]: if we're just asserting the percentage value here  (perecentText) maybe we could just assert "55%" since we don't have NumberUtils tests

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
```
Comment (luoser) [inline app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java]: should this be testing the pledged amount rather than percent text again?

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+      fifteenPercentExternalReferrer,
+      tenPercentExternalReferrer,
+      thirtyPercentInternalReferrer,
+      fiftyFivePercentCustomReferrer);
+    setUpEnvironment(environment());
+    final ProjectStatsEnvelope statsEnvelope = ProjectStatsEnvelopeFactory.projectStatsEnvelope().toBuilder().referralDistribution(referrerList).build();
+    this.vm.inputs.projectAndStatsInput(Pair.create(project, statsEnvelope));
+    this.customReferrerPercent.assertValues(55.0f);
+    this.externalReferrerPercent.assertValues(25.0f);
+    this.internalReferrerPercent.assertValues(30.0f);
```
Comment (luoser) [inline app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java]: great test!!

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java (hunk 7 above)
Comment (swoopej) [reply app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: oh lol tru

File: app/src/main/java/com/kickstarter/ui/viewholders/CreatorDashboardReferrerBreakDownViewHolder.java [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+
+    this.viewModel.outputs.internalReferrerPercent()
+      .compose(bindToLifecycle())
+      .compose(observeForUI())
+      .subscribe(percent -> this.referrerBreakdownView.setInternalAngleAndColor(percent * 360d));
+
+    this.viewModel.outputs.internalReferrerPercent()
+      .compose(bindToLifecycle())
+      .compose(observeForUI())
+      .subscribe(percent -> this.percentInternalTextView.setText(String.valueOf(percent * 100d)));
```
[the remark this answers is not in the document]
Comment (swoopej) [reply app/src/main/java/com/kickstarter/ui/viewholders/CreatorDashboardReferrerBreakDownViewHolder.java]: oooo good catch this is actually just unnecessary use of that output - we have `internalReferrerPercentText` below that does this for us. ty ty

File: app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java (hunk 16 above)
Comment (swoopej) [reply app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java]: ah yep bad copypaste

Comment (swoopej) [reply app/src/test/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModelTest.java]: also this is an unnecessary output so just going to remove from the viewmodel

File: app/src/main/java/com/kickstarter/ui/views/ReferrerBreakdownView.java [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    this.paint.setAntiAlias(true);
+    this.paint.setStyle(Paint.Style.FILL);
+
+    final float bottom = getHeight();
+    final float left = getX();
+    final float right = getWidth();
+    final float top = getY();
+
+    this.outerRectangle.set(left, top, right, bottom);
+    this.innerRectangle.set(left + 30, top + 30, right - 30, bottom - 30);
```
[the remark this answers is not in the document]
Comment (swoopej) [reply app/src/main/java/com/kickstarter/ui/views/ReferrerBreakdownView.java]: 👌 yep

**Comment by swoopej** [comment]
@luoser @eoji think I addressed all yalls feedback if you would take a look when you have a few mins.  Also now including a [fresh new drawable](https://github.com/kickstarter/android-oss/pull/168/files#diff-5bb5c8f7da918e47acc98b4a10d5971c) instead of the bullet text.

bitmoji

File: app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+      android:orientation="vertical">
+
+      <LinearLayout
+        android:id="@+id/pledged_via_kickstarter_layout"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:orientation="horizontal"
+        android:layout_weight="1">
+
+        <TextView
```
Comment (eoji) [inline app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml]: Since we're no longer using text, we can make this an `ImageView`

File: app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml [hunk 21]
```diff
@@ -0,0 +1,202 @@
...
+      </LinearLayout>
+
+      <LinearLayout
+        android:id="@+id/pledged_via_external_layout"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:orientation="horizontal"
+        android:layout_weight="1">
+
+        <TextView
```
Comment (eoji) [inline app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml]: Same here, we can make this an `ImageView`

File: app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+
+      <LinearLayout
+        android:id="@+id/pledged_via_custom_layout"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:orientation="horizontal"
+        android:layout_weight="1">
+
+
+        <TextView
```
Comment (eoji) [inline app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml]: One more!

File: app/src/main/java/com/kickstarter/ui/viewholders/CreatorDashboardReferrerBreakdownViewHolder.java [hunk 23]
```diff
@@ -147,4 +149,9 @@ private void setAveragePledgeTextViewText(final @NonNull Pair<Project, Integer>
     final String amountString = this.ksCurrency.format(projectAndAveragePledge.second, projectAndAveragePledge.first, true, true, RoundingMode.DOWN);
     this.averagePledgeAmountTextView.setText(amountString);
   }
+
+  private void setDrawableColor(final @NonNull TextView view, final @NonNull Integer color) {
```
Comment (eoji) [inline app/src/main/java/com/kickstarter/ui/viewholders/CreatorDashboardReferrerBreakdownViewHolder.java]: Since we'll be switching to an `ImageView`, we can change the color by doing `imageView.setColorFilter(color)`

Comment (swoopej) [reply app/src/main/java/com/kickstarter/ui/viewholders/CreatorDashboardReferrerBreakdownViewHolder.java]: oooo yes good point ty ty

Comment (swoopej) [reply app/src/main/java/com/kickstarter/ui/viewholders/CreatorDashboardReferrerBreakdownViewHolder.java]: and we can go back to point free 👯

File: app/src/main/res/drawable/creator_dashboard_circle.xml [hunk 24]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+  android:shape="oval">
+  <solid
+    android:color="@android:color/white"/>
```
Comment (eoji) [inline app/src/main/res/drawable/creator_dashboard_circle.xml]: Let's keep this transparent in case we evert decide to use a background color other than white.

File: app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml [hunk 25]
```diff
@@ -0,0 +1,205 @@
...
+      <LinearLayout
+        android:id="@+id/pledged_via_kickstarter_layout"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:orientation="horizontal"
+        android:layout_weight="1">
+
+        <ImageView
+          android:id="@+id/percent_via_kickstarter_circle_text_view"
+          android:contentDescription="@string/dashboard_graphs_referrers_pledged_via_kickstarter"
```
Comment (swoopej) [inline app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml]: @luoser wanted to make sure this was finesince I haven't really dealt with accessibility on android before.  These are just the bullets next to the graph but since they are image views, the linter wants them to have a description.  I'm just reusing the strings that we use in the textviews below for the description.

Comment (luoser) [reply app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml]: ah yes, the move here would actually to ignore the `contentDescription` since the circle imageview is just a visual indicator that doesn't give the user any new / meaningful information that the textviews provide

```xml
tools:ignore="ContentDescription"
```

should appease the linter!

Comment (swoopej) [reply app/src/main/res/layout/dashboard_referrer_breakdown_layout.xml]: ah amazing ty

File: app/src/main/java/com/kickstarter/ui/views/ReferrerBreakdownView.java [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+  public Canvas canvas;
+  private Double customSweepAngle;
+  private Double externalSweepAngle;
+  private Double internalSweepAngle;
+  private RectF outerRectangle;
+  private RectF innerRectangle;
+  private Paint paint;
+  private static final int innerCircleOffset = 30;
+
+  public ReferrerBreakdownView(final Context context, final AttributeSet attributeSet) {
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/ui/views/ReferrerBreakdownView.java]: can annotate params with `@NonNull` , as well as L45

File: app/src/main/java/com/kickstarter/ui/views/ReferrerBreakdownView.java [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+import android.graphics.Color;
+import android.graphics.Paint;
+import android.graphics.RectF;
+import android.support.annotation.NonNull;
+import android.util.AttributeSet;
+import android.view.View;
+
+import com.kickstarter.libs.ReferrerColor;
+
+public class ReferrerBreakdownView extends View {
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/ui/views/ReferrerBreakdownView.java]: class can be final

File: app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java [hunk 28]
```diff
@@ -0,0 +1,296 @@
...
+    Observable<Pair<Project, Integer>> projectAndAveragePledge();
+
+    /** Emits the current project and the amount pledged via custom referrers **/
+    Observable<Pair<Project, Float>> projectAndCustomReferrerPledgedAmount();
+
+    /** Emits the current project and the amount pledged via external referrers **/
+    Observable<Pair<Project, Float>> projectAndExternalReferrerPledgedAmount();
+
+    /** Emits the current project and the amount pledged via internal referrers **/
+    Observable<Pair<Project, Float>> projectAndInternalReferrerPledgedAmount();
```
Comment (luoser) [inline app/src/main/java/com/kickstarter/viewmodels/CreatorDashboardReferrerBreakdownHolderViewModel.java]: wahoo ty for the javadox
